sklearn/tree/tests/test_monotonic_tree.py

```diff
@@ -0,0 +1,348 @@
+                max_depth=None,
+                monotonic_cst=monotonic_cst,
+                max_leaf_nodes=n_samples_train,
+            )
+        if hasattr(est, "random_state"):
+            est.set_params(random_state=global_random_seed)
+        if hasattr(est, "n_estimators"):
+            est.set_params(**{"n_estimators": 5})
+        est.fit(X_train, y_train)
+        y = est.predict(X_test)
+        # increasing constraint
+        y_incr = est.predict(X_test_incr)
+        # y_incr should always be greater than y
+        assert np.all(y_incr >= y)
+
+        # decreasing constraint
+        y_decr = est.predict(X_test_decr)
+        # y_decr should always be lower than y
+        assert np.all(y_decr <= y)
+
+
+def test_multiclass_raises():
+    X, y = make_classification(
+        n_samples=100, n_features=5, n_classes=3, n_informative=3, random_state=0
+    )
+    y[0] = 0
+    monotonic_cst = np.zeros(X.shape[1])
+    monotonic_cst[0] = -1
+    monotonic_cst[1] = 1
+    for name, TreeClassifier in CLF_TREES.items():
+        est = TreeClassifier(
+            max_depth=None, monotonic_cst=monotonic_cst, random_state=0
+        )
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": 0})
+
+        msg = "Monotonic constraints are not supported with multiclass classification"
+        with pytest.raises(ValueError, match=msg):
+            est.fit(X, y)
+
+
+def test_multiple_output_raises():
+    X = [[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]]
+    y = [[1, 0, 1, 0, 1], [1, 0, 1, 0, 1]]
+
+    for name, TreeClassifier in CLF_TREES.items():
+        est = TreeClassifier(
+            max_depth=None, monotonic_cst=np.array([-1, 1]), random_state=0
+        )
+        msg = "Monotonic constraints are not supported with multiple output"
+        with pytest.raises(ValueError, match=msg):
+            est.fit(X, y)
+
+
+def test_bad_monotonic_cst_raises():
+    X = [[1, 2], [3, 4], [5, 6], [7, 8], [9, 10]]
+    y = [1, 0, 1, 0, 1]
+
+    for name, TreeClassifier in CLF_TREES.items():
+        msg = "monotonic_cst has shape 3 but the input data X has 2 features."
+        est = TreeClassifier(
+            max_depth=None, monotonic_cst=np.array([-1, 1, 0]), random_state=0
+        )
+        with pytest.raises(ValueError, match=msg):
+            est.fit(X, y)
+
+        msg = "monotonic_cst must be None or an array-like of -1, 0 or 1."
+        est = TreeClassifier(
+            max_depth=None, monotonic_cst=np.array([-2, 2]), random_state=0
+        )
+        with pytest.raises(ValueError, match=msg):
+            est.fit(X, y)
+
+        est = TreeClassifier(
+            max_depth=None, monotonic_cst=np.array([-1, 0.8]), random_state=0
+        )
+        with pytest.raises(ValueError, match=msg):
+            est.fit(X, y)
+
+
+def assert_1d_reg_tree_children_monotonic_bounded(tree_, monotonic_sign):
+    # Flip values to always check for increasing constraint
+    values = monotonic_sign * tree_.value
```

Comment:
I think the logic doesn't work this way and we need a case distinction instead:
```python
if increasing.
    assert values[i_left] <= values[i_right]
else:
    assert values[i_left] >= values[i_right]

Reply:
I think both are equivalent but what you propose is more explicit and understandable.

Reply:
I also agree that the suggestion above would make the test easier to grasp.

Reply:
Done in 55a05444 too.